File: ProblemSolving/450DSA/JavaScript/src/binarytree/PreorderTraversal.js
/**
 * @author Anirudh Sharma
 *
 * Given a binary tree, write an iterative and recursive solution to traverse the tree using preorder traversal.
 */
const preorderTraversalRecursively = (root) => {
    if (root === null) {
        return null;
    }
    console.log(root.data);
    preorderTraversalRecursively(root.left);
    preorderTraversalRecursively(root.right);
};

const preorderTraversalIteratively = (root) => {
    if (root === null) {
        return null;
    }
    // Stack to store the nodes of the tree
    const nodes = [];
    // Add root to the stack
    nodes.push(root);
    // Loop until the stack is not empty
    while (nodes.length > 0) {
        // Get the current node from the stack
        let current = nodes.pop();
        console.log(current.data);
        // Push right child of the tree to stack if not null
        if (current.right !== null) {
            nodes.push(current.right);
        }
        // Push left child of the tree to stack if not null
        if (current.left !== null) {
            nodes.push(current.left);
        }
    }
};

function Node(data, left, right) {
    this.data = (data === undefined ? 0 : data);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

const main = () => {
    let root = new Node(1);
    root.left = new Node(3);
    root.right = new Node(2);
    console.log("Recursively:");
    preorderTraversalRecursively(root);
    console.log("Iteratively:");
    preorderTraversalIteratively(root);
    console.log();

    root = new Node(10);
    root.left = new Node(20);
    root.right = new Node(30);
    root.left.left = new Node(40);
    root.left.right = new Node(60);
    console.log("Recursively:");
    preorderTraversalRecursively(root);
    console.log("Iteratively:");
    preorderTraversalIteratively(root);
    console.log();

    root = new Node(1);
    root.left = new Node(2);
    root.right = new Node(3);
    root.left.left = new Node(4);
    root.left.right = new Node(5);
    console.log("Recursively:");
    preorderTraversalRecursively(root);
    console.log("Iteratively:");
    preorderTraversalIteratively(root);
    console.log();
};

main();

File: ProblemSolving/450DSA/Python/src/binarytree/PreorderTraversal.py
"""
@author Anirudh Sharma

Given a binary tree, write an iterative and recursive solution to traverse the tree using preorder traversal.
"""


def preorderTraversalRecursively(root):
    # Base case
    if root is None:
        return
    print(root.data, end=" ")
    preorderTraversalRecursively(root.left)
    preorderTraversalRecursively(root.right)


def preorderTraversalIteratively(root):
    if root is None:
        return None
    # Stack to store the nodes of the tree
    nodes = []
    # Add root node to the stack
    nodes.append(root)
    # Loop until stack is empty
    while nodes:
        # Get current node
        current = nodes.pop()
        print(current.data, end=" ")
        # Push right child if not null
        if current.right:
            nodes.append(current.right)
        # Push left child if not null
        if current.left:
            nodes.append(current.left)


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(3)
    root.right = Node(2)
    print("Recursive:")
    preorderTraversalRecursively(root)
    print("\nIterative:")
    preorderTraversalIteratively(root)
    print("\n")

    root = Node(10)
    root.left = Node(20)
    root.right = Node(30)
    root.left.left = Node(40)
    root.left.right = Node(60)
    print("Recursive:")
    preorderTraversalRecursively(root)
    print("\nIterative:")
    preorderTraversalIteratively(root)
    print("\n")

    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    print("Recursive:")
    preorderTraversalRecursively(root)
    print("\nIterative:")
    preorderTraversalIteratively(root)
    print("\n")
